src/tribol/physics/Mortar.cpp

```diff
@@ -631,6 +641,565 @@ void ComputeSingleMortarJacobian( SurfaceContactElem& elem )
+      xq_guess[0] += phiq[i] * xq[i];
+      xq_guess[1] += phiq[i] * yq[i];
+    }
+    RealT resid[2];
+    resid[0] = xq_guess[0] - xt[0];
+    resid[1] = xq_guess[1] - xt[1];
+    RealT max_resid = std::max( std::abs( resid[0] ), std::abs( resid[1] ) );
+    if ( max_resid < conv ) {
+      break;
+    }
+    ++iter_ct;
+    if ( iter_ct > max_iter ) {
+      std::cout << "Max iteration count reached.  Max resid = " << max_resid << std::endl;
+      break;
+    }
+    // update xiq
+    RealT dphiq[8];
+    LinIsoQuadShapeFuncDeriv( xiq, dphiq );
+    RealT dresid_dxiq[4] = { 0.0, 0.0, 0.0, 0.0 };
+    for ( int i{ 0 }; i < 4; ++i ) {
+      dresid_dxiq[0] += dphiq[i] * xq[i];
+      dresid_dxiq[1] += dphiq[i] * yq[i];
+      dresid_dxiq[2] += dphiq[4 + i] * xq[i];
+      dresid_dxiq[3] += dphiq[4 + i] * yq[i];
+    }
+    // explicit inverse
+    RealT detJ = dresid_dxiq[0] * dresid_dxiq[3] - dresid_dxiq[2] * dresid_dxiq[1];
+    xiq[0] -= ( dresid_dxiq[3] * resid[0] - dresid_dxiq[2] * resid[1] ) / detJ;
+    xiq[1] -= ( -dresid_dxiq[1] * resid[0] + dresid_dxiq[0] * resid[1] ) / detJ;
+  }
+}
+
+//------------------------------------------------------------------------------
+void ComputeMortarForceEnzyme( const RealT* x1, const RealT* n1, const RealT* p1, RealT* f1, RealT* g1, int size1,
+                               const RealT* x2, RealT* f2, int size2 )
+{
+  // convention: elem1 = nonmortar element
+  //             elem2 = mortar element
+  constexpr int max_mortar_mat_size = 4 * 4;
+  RealT mortar_mat1[max_mortar_mat_size];
+  int mortar_mat1_size = size1 * size1;
+  for ( int i{ 0 }; i < mortar_mat1_size; ++i ) {
+    mortar_mat1[i] = 0.0;
+  }
+  RealT mortar_mat2[max_mortar_mat_size];
+  int mortar_mat2_size = size1 * size2;
+  for ( int i{ 0 }; i < mortar_mat2_size; ++i ) {
+    mortar_mat2[i] = 0.0;
+  }
+  // get point x0 (geometric center of elem1)
+  RealT x0[3] = { 0.0, 0.0, 0.0 };
+  for ( int i{ 0 }; i < size1; ++i ) {
+    for ( int d{ 0 }; d < 3; ++d ) {
+      x0[d] += x1[d * size1 + i] / static_cast<RealT>( size1 );
+    }
+  }
+
+  // get vector n (normal of elem1)
+  // NOTE: this limits this routine to quads
+  // clang-format off
+   RealT de1[3] = {
+      -0.25*x1[0] + 0.25*x1[1] + 0.25*x1[2] - 0.25*x1[3],
+      -0.25*x1[4] + 0.25*x1[5] + 0.25*x1[6] - 0.25*x1[7],
+      -0.25*x1[8] + 0.25*x1[9] + 0.25*x1[10] - 0.25*x1[11]
+   };
+   RealT de2[3] = {
+      -0.25*x1[0] - 0.25*x1[1] + 0.25*x1[2] + 0.25*x1[3],
+      -0.25*x1[4] - 0.25*x1[5] + 0.25*x1[6] + 0.25*x1[7],
+      -0.25*x1[8] - 0.25*x1[9] + 0.25*x1[10] + 0.25*x1[11]
+   };
+   RealT n[3] = {
+      de1[1]*de2[2] - de1[2]*de2[1],
+      de1[2]*de2[0] - de1[0]*de2[2],
+      de1[0]*de2[1] - de1[1]*de2[0]
+   };
+  // clang-format on
+  RealT n_mag = std::sqrt( n[0] * n[0] + n[1] * n[1] + n[2] * n[2] );
+  for ( int d{ 0 }; d < 3; ++d ) {
+    n[d] /= n_mag;
+  }
+
+  // x1t = x1 projected to plane p (def'd by x0 and n)
```

Comment:
Comment is a little cryptic. I don't know what x1t is.

Reply:
added